lib/Doctrine/ORM/Configuration.php

```diff
@@ -918,3 +919,5 @@ public function setDefaultQueryHint($name, $value)
         $this->_attributes['defaultQueryHints'][$name] = $value;
     }
 }
+
+interface_exists(MappingDriver::class);
```

Comment:
This is a side-effect of a non-optimal decision done in the upstream that I'd love to avoid by all means necessary - yes I read the blog post and I disagree with the approach.

Reply:
I think it could be avoided by bumping the version of php to something that does not require that piece of black magic… I didn't manage to figure out which one work yet though.

Reply:
Do you disagree with the whole approach or just the last part? If you know of other approaches, please share them.

Reply:
> This is a side-effect of a non-optimal decision done in the upstream that I'd love to avoid by all means necessary

The team a while back decided to go ahead with the namespace change from `Doctrine\Common\Persistence` to `Doctrine\Persistence`, against the advice from a few people warning of the impact of such a change (not specifically for persistence, but for annotations where the impact is pretty much the same, see also https://github.com/doctrine/annotations/pull/75#discussion_r56872787). This polyfill layer is the only option we have to avoid a hard BC break that will either keep people from upgrading to v2 or other projects (e.g. Symfony) having to maintain their own compatibility layers (which previously caused protests from other library authors).

So while you may disagree with the approach (as others have done), there has yet to be a specific suggestion that allows for such a significant change in a widely used library without the need for this kind of glue code. If you have a suggestion, by all means, let us know. Until then, this is the way this will have to be done.

Reply:
@lcobucci as I said at the end of my blog post, I'm not a huge fan of this approach either: it's ugly. But IMO it's far better than having no BC-layer at all. I think there should be way more of them in Doctrine and I'm afraid several Doctrine packages are going to be quite a pain to migrate to their next major version. That being said, I don't know that much about the differences between branches in the ORM, for instance, so I hope I'm wrong.

Reply:
I agree on the need for a BC-layer (no need for discussions there).

The problem is the usage of class/interface aliases, having explicit interfaces/classes extending the new stuff is much cleaner, is cacheable by opcache (and preoloaded properly), doesn't force us to have these calls, and doesn't break anything on 3rd party.

Reply:
Note that class aliases *are* compatible with opcache/preloading.

Reply:
> having explicit interfaces/classes extending the new stuff is much cleaner

Ok so let's say you have `LegacyFoo extends NewShinyFoo`.

If you have an interface that has a method with a type declaration, and you change it to `NewShinyFoo`, you have a BC break, because any implementor will restrict the type to `LegacyFoo` instead of widening it, which is possible in theory but only in php 7.4

Reply:
@nicolas-grekas not according to the RFC:

> Only classes without unresolved parent, interfaces, traits and constant values may be preloaded. If a class doesn't satisfy to this condition, it's stored in opcache SHM as a part of corresponding PHP script in the same way as without preloading. Also, only top-level entities that are not nested within control structures (e.g. if()…) may be preloaded.

I get a `Notice` when trying to preload stuff, like `[16-Dec-2019 16:53:42 UTC] PHP Warning:  Can't preload unlinked class TestC: Unknown interface TestB in /app/C.php on line 3`. Please let me know if I'm doing something wrong.

<details>
  <summary>Reproducible script</summary>

*A.php*
```php
<?php

interface TestA
{
}

class_alias(TestA::class, TestB::class);
``` 

*C.php*
```php
<?php

final class TestC implements TestB
{
    public function run(): void
    {
        echo 'blah', PHP_EOL;
    }
}
```

*D.php*
```php
<?php

require 'A.php';
require 'C.php';

(new TestC())->run();
```

*preload.php*
```php
<?php

if (! opcache_compile_file(__DIR__ . '/A.php')) {
    trigger_error("Preloading Failed", E_USER_ERROR);
}

if (! opcache_compile_file(__DIR__ . '/C.php')) {
    trigger_error("Preloading Failed", E_USER_ERROR);
}
```

*Dockerfile*
```Dockerfile
FROM php:7.4-alpine

WORKDIR /app

RUN { \
    echo 'zend.assertions=-1'; \
    echo 'assert.exception=1'; \
    echo 'opcache.enable_cli=1'; \
    echo 'opcache.max_wasted_percentage=10'; \
    echo 'opcache.validate_timestamps=0'; \
    echo 'opcache.enable_file_override=1'; \
    echo 'opcache.optimization_level=0xFFFFFFFF'; \
    echo 'opcache.preload=/app/preload.php'; \
    echo 'opcache.preload_user=root'; \
    echo 'error_reporting=E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED & ~E_STRICT'; \
    echo 'display_errors=1'; \
    echo 'log_errors=1'; \
    echo 'error_log=/dev/stderr'; \
} | tee /usr/local/etc/php/conf.d/10-prod.ini

RUN apk add --no-cache ${PHPIZE_DEPENDENCIES} \
    && docker-php-ext-install opcache

COPY A.php C.php D.php preload.php ./

CMD ["php", "D.php"]
```

Run:

```bash
$ docker build -t preload-test . && docker run --rm preload-test
```
</details>

Reply:
The commit is quite new:
https://github.com/php/php-src/commit/baf3a9133bf0bc7069e5191b780c221485b2dda6

If you already have a recent enough build, maybe it works with "require" but not with "opcache_compile_file()"?

Reply:
@nicolas-grekas that targets PHP 7.4.1, so we don't have something that works with stable the version. Good to know it's being addressed, though.

Reply:
[Nikita confirmed that `class_alias()` will never work with `opcache_compile_file()`](https://github.com/php/php-src/commit/baf3a9133bf0bc7069e5191b780c221485b2dda6#commitcomment-36450273). Using `require` isn't an option since it executes the code, which is quite undesired.

Anyhow, thanks for your input @nicolas-grekas.

Reply:
> Nikita confirmed that class_alias() will never work with opcache_compile_file().

Great, thanks for the hint. I think this can be resolved in the future - opcache is already able to resolve some functions at compiler time. I'm sure it will happen for class_alias() if it's worth it at some point :)

> Using require isn't an option since it executes the code, which is quite undesired.

But you have zero control over this (you = any lib author). Actually, "require" is the preferred way of configuring preloading in Symfony, *because* it is executing the code, thus it works with conditional classes or similar exotic situations that happen in real life.

Reply:
> But you have zero control over this (you = any lib author)

You're completely correct that I don't have control over what users do. However, I do have control if the code I produce is fully compatible with `opcache_compile_file()`.

> Actually, "require" is the preferred way of configuring preloading in Symfony, because it is executing the code, thus it works with conditional classes or similar exotic situations that happen in real life.

Yeah... I rather keep the preoloading of my apps clean of execution, speeding up such process and avoiding unintended side-effects. Especially because is the intended way for as far as the RFC tells (but maybe @nikic or @dstogov can present arguments disproving my understanding).

@nicolas-grekas thanks sharing your PoV (despite our disagreement) :+1:

Reply:
@lcobucci I think at this point I would recommend the require model. The opcache_compile_file() variant may stop being tenable as projects add property types (and thus introduce circular dependencies). I'm not sure at this point if it's worth investing in making that work (quite hard) when the require based approach can handle this and more transparently.

Reply:
For me this is a good sign that the namespace changes away from Common don't add value, but only problems. Its unfortunate that we have to deal with this stuff, but we can still roll back on it to simplify our lives.

Reply:
They do add a little value IMO: it's now easier to troubleshoot what package is involved when getting error messages that include that kind of FQCN. It's less confusing. Shorter namespaces also help fuzzy matchers perform better when finding types. And it's not like this BC layer should stay here forever.